File: packages/hydra-java/src/main/java/hydra/dsl/meta/lib/Optionals.java
package hydra.dsl.meta.lib;
import hydra.dsl.meta.Phantoms;
import hydra.typed.TypedTerm;

/**
 * Phantom-typed term DSL for the {@code hydra.lib.optionals} library.
 *
 * <p>Java analogue of {@code hydra.dsl.meta.lib.optionals} in Python and
 * {@code Hydra.Dsl.Meta.Lib.Optionals} in Haskell.</p>
 *
 * <p>Each method here builds an {@code apply}-chain calling the corresponding
 * primitive function (e.g. {@code hydra.lib.optionals.fromOptional}).</p>
 */
public final class Optionals {
    private Optionals() {}

    /** Reference to {@code hydra.lib.optionals.<name>}. */
    private static <A> TypedTerm<A> prim(String name) {
        return Phantoms.var("hydra.lib.optionals." + name);
    }

    /** {@code Optionals.apply f x} — apply a function in Maybe context (applicative). */
    public static <A> TypedTerm<A> apply(TypedTerm<?> f, TypedTerm<?> x) {
        return Phantoms.apply(prim("apply"), f, x);
    }

    /** {@code Optionals.bind x f} — chain operations on optional values. */
    public static <A> TypedTerm<A> bind(TypedTerm<?> x, TypedTerm<?> f) {
        return Phantoms.apply(prim("bind"), x, f);
    }

    /** {@code Optionals.cases m default just_handler}. */
    public static <A> TypedTerm<A> cases(TypedTerm<?> m, TypedTerm<?> defaultBranch, TypedTerm<?> justBranch) {
        return Phantoms.apply(prim("cases"), m, defaultBranch, justBranch);
    }

    /** {@code Optionals.cat xs} — keep only Just values from a list. */
    public static <A> TypedTerm<A> cat(TypedTerm<?> xs) {
        return Phantoms.apply(prim("cat"), xs);
    }

    /** {@code Optionals.compose f g}. */
    public static <A> TypedTerm<A> compose(TypedTerm<?> f, TypedTerm<?> g) {
        return Phantoms.apply(prim("compose"), f, g);
    }

    /** {@code Optionals.fromJust x} — extract value (unsafe if Nothing). */
    public static <A> TypedTerm<A> fromJust(TypedTerm<?> x) {
        return Phantoms.apply(prim("fromJust"), x);
    }

    /** {@code Optionals.fromOptional default x} — get the value or default. */
    public static <A> TypedTerm<A> fromOptional(TypedTerm<?> defaultVal, TypedTerm<?> x) {
        return Phantoms.apply(prim("fromOptional"), defaultVal, x);
    }

    /** {@code Optionals.isGiven x}. */
    public static <A> TypedTerm<A> isGiven(TypedTerm<?> x) {
        return Phantoms.apply(prim("isGiven"), x);
    }

    /** {@code Optionals.isNone x}. */
    public static <A> TypedTerm<A> isNone(TypedTerm<?> x) {
        return Phantoms.apply(prim("isNone"), x);
    }

    /** {@code Optionals.map f x}. */
    public static <A> TypedTerm<A> map(TypedTerm<?> f, TypedTerm<?> x) {
        return Phantoms.apply(prim("map"), f, x);
    }

    /** {@code Optionals.mapOptional f xs}. */
    public static <A> TypedTerm<A> mapOptional(TypedTerm<?> f, TypedTerm<?> xs) {
        return Phantoms.apply(prim("mapOptional"), f, xs);
    }

    /** {@code Optionals.pure x} — wrap in Just. */
    public static <A> TypedTerm<A> pure(TypedTerm<?> x) {
        return Phantoms.apply(prim("pure"), x);
    }

    /** {@code Optionals.toList x} — Just x → [x], Nothing → []. */
    public static <A> TypedTerm<A> toList(TypedTerm<?> x) {
        return Phantoms.apply(prim("toList"), x);
    }
}

File: heads/haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Optionals.hs
{-# LANGUAGE FlexibleContexts #-}

-- | Phantom-typed term DSL for the hydra.lib.optionals library

module Hydra.Dsl.Meta.Lib.Optionals where

import Hydra.Typed
import Hydra.Dsl.AsTerm
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries


-- | Apply a function to an argument (applicative).
apply :: TypedTerm (Maybe (a -> b)) -> TypedTerm (Maybe a) -> TypedTerm (Maybe b)
apply = primitive2 _optionals_apply

-- | Chain operations on optional values, handling absent cases automatically.
bind :: TypedTerm (Maybe a) -> TypedTerm (a -> Maybe b) -> TypedTerm (Maybe b)
bind = primitive2 _optionals_bind

-- | Handle an optional value with the optional value as the first argument.
-- The default and function arguments accept anything coercible to a term (AsTerm),
-- mirroring the ergonomics of the former 'maybe' eliminator.
cases :: (AsTerm t1 b, AsTerm t2 (a -> b)) => TypedTerm (Maybe a) -> t1 -> t2 -> TypedTerm b
cases m def f = primitive3 _optionals_cases m (asTerm def) (asTerm f)

-- | Filter out absent values from a list.
cat :: TypedTerm [Maybe a] -> TypedTerm [a]
cat = primitive1 _optionals_cat

-- | Compose two optional-returning functions (Kleisli composition).
compose :: TypedTerm (a -> Maybe b) -> TypedTerm (b -> Maybe c) -> TypedTerm (a -> Maybe c)
compose = primitive2 _optionals_compose

-- | Get a value from an optional value, or return a default value.
fromOptional :: TypedTerm a -> TypedTerm (Maybe a) -> TypedTerm a
fromOptional = primitive2 _optionals_fromOptional

-- | Check if a value is present.
isGiven :: TypedTerm (Maybe a) -> TypedTerm Bool
isGiven = primitive1 _optionals_isGiven

-- | Check if a value is absent.
isNone :: TypedTerm (Maybe a) -> TypedTerm Bool
isNone = primitive1 _optionals_isNone

-- | Map a function over an optional value.
map :: AsTerm f (a -> b) => f -> TypedTerm (Maybe a) -> TypedTerm (Maybe b)
map f = primitive2 _optionals_map (asTerm f)

-- | Map a function over a list and collect present results.
mapOptional :: TypedTerm (a -> Maybe b) -> TypedTerm [a] -> TypedTerm [b]
mapOptional = primitive2 _optionals_mapOptional

-- | Lift a value into the optional type.
pure :: TypedTerm a -> TypedTerm (Maybe a)
pure = primitive1 _optionals_pure

-- | Convert an optional value to a list: a present value becomes [x], an absent value becomes [].
toList :: TypedTerm (Maybe a) -> TypedTerm [a]
toList = primitive1 _optionals_toList
